### PR TITLE
Add read-ahead

### PR DESCRIPTION
**Pull request by tomCynnySpace**

### CODE REVIEW

**Comment by tomCynnySpace** [comment]
sure,
I can do some other changes, we want a rock solid S3QL! :)

File: src/s3ql/block_cache.py [hunk 1]
```diff
@@ -272,6 +273,7 @@ def __init__(self, backend_pool, db, cachedir, max_size, max_entries=768):
         self.upload_threads = []
         self.removal_threads = []
         self.transfer_completed = SimpleEvent()
+        self.read_ahead = ReadAhead()
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/block_cache.py]: done

File: src/s3ql/block_cache.py [hunk 2]
```diff
@@ -755,6 +757,31 @@ def get(self, inode, blockno):
         if self.cache.is_full():
             self.expire()
 
+        if inode not in self.read_ahead.ReadAheadData.keys() or self.read_ahead.ReadAheadData[inode].Expired(): # new file, fill the read_ahead data structure
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/block_cache.py]: done

File: src/s3ql/block_cache.py [hunk 3]
```diff
@@ -755,6 +757,31 @@ def get(self, inode, blockno):
...
+          try:
+            list_block_no = self.db.get_list('SELECT DISTINCT blockno FROM inode_blocks WHERE inode=?', (inode,))
+            log.debug('inode =%d, list block_no=%s', inode, list_block_no)
+            dict_blockno_objid={}
+        
+            for bn in list_block_no:
+                bn=bn[0] # de-tuple the value
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/block_cache.py]: done

File: src/s3ql/read_ahead.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+'''
+read_ahead.py - this file is part of S3QL.
+
+Copyright  2018 Tommaso Massimi <t.massim@cynnyspace.com>
+
+This work can be distributed under the terms of the GNU GPLv3.
+'''
+
+
+import threading, time, os, shutil
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/read_ahead.py]: done

File: src/s3ql/read_ahead.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  class logging():
+    def debug(a,*args):
+      print(args[0])
+  
+  log = logging()
+
+
+class ReadAhead_details():
+    '''
+    keeps read_ahead variables for each file
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/read_ahead.py]: done

File: src/s3ql/read_ahead.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    '''
+    def __init__(self, dict_block_no_objid, blockno):
+        self.last_block_read = blockno
+        self.consec_block_count = 1
+        self.RAlast_access = time.time()
+
+        self.dict_block_no_objid = dict_block_no_objid
+        try:
+            del self.dict_block_no_objid[blockno]        # remove this block from the list of blocks to be downloaded
```
[the remark this answers is not in the document]
Comment (tomCynnySpace) [reply src/s3ql/read_ahead.py]: let suppose we are downloading a file.
the syncronous code does these calls:

- GET(block=0) downlaods fileobj=x
- GET(block=1) downlaods fileobj=y
- GET(block=2) downlaods fileobj=z

during this last GET, the asyncronuos mechanism decide to start its own threads; 
each threads downloads a fileobj but the calling GET is ready to download fileobj=z.

this is the reason why I'm removing fileobj z from the list of fileobj to download

**Comment by Nikratio** [comment]
Friendly ping. What's the status of this PR? Will you be able to add tests and documentation?

**Comment by tomCynnySpace** [comment]
about tests: my code adds no new features, S3QL works like in the past but
it is faster in retrieving big files.  how can I add a test and bu sure
that the new code is used?

about documentation: I have added docstrings in the code, there are some
other pages where I should add comments?

there are still little improvement to do to this code, but we are using it
with good results, we have only some little trouble with the size of the
cache

**Comment by Nikratio** [comment]
The easiest way to test the new code is to write tests that call into it directly. So you would add tests that call the various new functions in `read_ahead.py` and confirm that they actually do what they should be doing. Then there should be an integration test that makes sure that the high-level feature is correct. It would read some data (triggering read-ahead) and then read some of the data that ought to have been pre-fetched. Prior to the second read, however, you have to make sure that the data has indeed been read previously. One way to do that is to remove the storage object from underneath S3QL - if prefetch didn't work, the second read would then fail when attempting to read from the backend.

In terms of documentation, this should be documented in the manual probably in `rst/mount.rst`. Users should be able to tell roughly how read-head works, so that they can decide if it would help in their scenario. For example, I would like to know if every read triggers a read-ahead, or only a number of consecutive reads, and how many blocks S3QL will read ahead. I'd also want to know how this interacts with caching - when will read-ahead blocks be removed from the cache? Do they count as "just accessed" (i.e., they will be removed last) or as "never accessed" (i.e., they will be removed first).

File: src/s3ql/block_cache.py [hunk 7]
```diff
@@ -797,33 +826,51 @@ def _get_entry(self, inode, blockno):
...
+            # check and start asynchronous downloads
+            self.read_ahead.block_download_required(self.path, inode, blockno, obj_id, self.backend_pool)
+
+            log.debug('block_download_required finished: inode=%d, blockno=%d, obj_id=%d' % (inode, blockno, obj_id))
+
+            # TODO if the block is in asynchronous download, this thread must wait
+            while os.path.isfile(filename + '.async_tmp'): #obj_id in self.read_ahead.listObjIDDwnInProgress:
+                log.debug('file %s present,, waiting for its end downlaod' % (filename + '.async_tmp'))
+                time.sleep(1000)
+                # adesso questo blocco potrebbe essere nella cache...andrebbe saltata la cache
```
Comment (andrewchambers) [inline src/s3ql/block_cache.py]: English please.

File: src/s3ql/read_ahead.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+Copyright  2018 Tommaso Massimi <t.massim@cynnyspace.com>
+
+This work can be distributed under the terms of the GNU GPLv3.
+'''
+
+"""
+ HOW THE READ AHEAD MECHANISM WORKS
+
+  during  a file uplaod, S3QL desects the original file in smaller files named
```
Comment (andrewchambers) [inline src/s3ql/read_ahead.py]: 'uplaod' 'downlaod'  'desects' typos.

File: src/s3ql/read_ahead.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  if a whole file is required, all the s3ql_data_$obj_id sections of the original file 
+  are retrieved in sequence, one after the other one
+
+  the READ AHEAD mechanism tries to speed up the download of a file:
+  when some consecutive sections of a file are required, 
+  it starts some threads which will download some s3ql_data_$obj_id sections 
+  of the file also if these section were not yet required
+
+  to do that the READ AHEAD mechanism needs to:
+   a) have the list of $obj_id of the donloaded file
```
Comment (andrewchambers) [inline src/s3ql/read_ahead.py]: 'donloaded' should be fixed.

**Comment by segator** [comment]
How the system detect that needs to download in readahead mode?
or always try to download next blocks?
So if I request for first 5MB of every file I will able download 2 Objects(guess block size is 10M)

**Comment by tomCynnySpace** [comment]
let suppose that you have a 200 MB file; you'll have 20 blocks of 10MB.

when you try to retrieve the 200MB file, you read the first byte and s3ql
downloads the first 10MB  block.

after a while, you require some bytes over 10MB of your 200MB file, and
s3ql downloads the second 10MB block.

after a while, you require some bytes over 20MB of your 200MB file, and
s3ql downloads the third 10MB block; in the meantime  the readahead
mechanism detects that you required 3 consecutive 10MB blocks so it decide
to download other blocks starting some threads, one for each 10MB block

after a while, you require some bytes over 30MB of your 200MB file, and
s3sl find the forth block already downloaded.



Il giorno sab 8 dic 2018, 19:14 Isaac Aymerich <notifications@github.com>
ha scritto:

**Comment by Nikratio** [comment]
For merging, this still needs tests and documentation (inside the source code, not in the pull request comments).

**Comment by Nikratio** [comment]
Friendly ping. Do you think you'll have time to resolve the open issues in the near future? If not, I'll close this pull request for now.

**Comment by Nikratio** [comment]
Closing this for now as discussed above. Please do reopen once the open issues are resolved!

**Pull request closed without merging**